poppy/accel_math.py

```diff
@@ -64,16 +88,220 @@ def _exp(x):
+        wavefront = _ifftshift(wavefront)
+
+    if _USE_CUDA:
+        if normalization is None:
+            normalization = 1./wavefront.shape[0]  # regardless of direction, for CUDA
+
+        # We need a CUDA FFT plan for each size and shape of FFT.
+        cufftplan = pyculib.fft.FFTPlan( wavefront.shape, wavefront.dtype, wavefront.dtype)
+
+        # The plans could in principle be cached for reuse
+        # FIXME - there appears to be a memory leak with doing it this way. Something never gets released
+        # when we try to cache this, eventually leading to GPU runtime failure from inability to malloc
+        # new array buffers. . The plan creation takes < 0.5 ms so it's perhaps acceptable to not cache 
+        # the results. At least for now this is worth it for the much better reliability
+        #try:
+            #params = (wavefront.shape, wavefront.dtype, wavefront.dtype)
+            #cufftplan = _CUDA_PLANS[params]
+            #print("cache hit, {},  {}".format(params, cufftplan))
+        #except KeyError:
+            #cufftplan = pyculib.fft.FFTPlan(*params)
+            #_CUDA_PLANS[params] = cufftplan
+            #print("cache miss, {},  {}".format(params, cufftplan))
+
+        # perform FFT on GPU, and return results in place to same array.
+        if forward:
+            cufftplan.forward(wavefront, out=wavefront)
+        else:
+            cufftplan.inverse(wavefront, out=wavefront)
+
+    elif _USE_OPENCL:
+        if normalization is None:
+            normalization = 1./wavefront.shape[0] if forward else wavefront.shape[0]
+
+        context, queue = get_opencl_context()
+        wf_on_gpu = pyopencl.array.to_device(queue, wavefront)
+        transform = gpyfft.fft.FFT(context, queue, wf_on_gpu, axes=(0,1))
+        event, = transform.enqueue(forward=forward)
+        event.wait()
+        wavefront[:] = wf_on_gpu.get()
+        del wf_on_gpu
+
+    elif _USE_FFTW:
+        FFT_direction = 'forward' if forward else 'backward' # back compatible for use in _FFTW_INIT
+        do_fft = pyfftw.interfaces.numpy_fft.fft2 if forward else pyfftw.interfaces.numpy_fft.ifft2
+        if normalization is None:
+            normalization = 1./wavefront.shape[0] if forward else wavefront.shape[0]
+
+
+        if (wavefront.shape, FFT_direction) not in _FFTW_INIT:
+            # The first time you run FFTW to transform a given size, it does a speed test to
+            # determine optimal algorithm that is destructive to your chosen array.
+            # So only do that test on a copy, not the real array:
+            _log.info("Measuring pyfftw optimal plan for %s, direction=%s" % (
+                str(wavefront.shape), FFT_direction))
+
+            pyfftw.interfaces.cache.enable()
+            pyfftw.interfaces.cache.set_keepalive_time(30)
+
+            test_array = np.zeros(wavefront.shape)
+            test_array = do_fft(test_array, overwrite_input=True, planner_effort='FFTW_MEASURE',
```

Comment:
I'm starting to suspect I may have been wrong before about there being a memory leak from storing the plans. It's possible the root cause was actually there's a memory leak _when test cases fail during pytest_, in which case exceptions are raised and variables do not get cleaned up properly.  In which case now that the tests are passing, that should not happen.  

I'm cautiously optimistic this may be the case, and am running the test suite repeatedly on my laptop with CUDA to see...

Reply:
I am increasingly convinced this is the case!